_posts/2019-12-29-working-with-interfaces-in-go.md

```diff
@@ -0,0 +1,140 @@
+    return &aesCipherGCM{c}, nil
+  }
+  // Returned type is aesCipherAsm.
+  return &c, nil
+}
+```
+
+Note that in the previous example, the interface was defined in the producer package `rand`. However in this example, the returned type is defined in a different package `cipher`.
+
+**In my experience ...**
+
+... this pattern is a lot more difficult to execute than the previous ones. In the early phases of development the needs of a consumer evolves quickly. Hence, it results in modifying the returned interface from the producer package. The interface slowly becomes too big to the point where returning a concrete type would have made more sense.
+
+My hypothesis for the mechanics of this pattern is:
+
+1. The interface returned needs to be small so that there can be multiple implementations.
+2. Hold off on returning an interface until you have multiple types in your package only implementing the interface. Multiple types gives confidence that you've the right interface.
+
+#### [observation] Consider creating a separate interfaces only package for namespacing and standardization
+
+Another common pattern in the standard library is having packages that contain only interfaces.
+
+An example is the [`hash.Hash`](https://golang.org/pkg/hash/) interface that’s implemented by the packages under the subdirectories of `hash/`  such as [`hash/crc32` ](https://golang.org/pkg/hash/crc32/) and [`hash/adler32`](https://golang.org/pkg/hash/adler32/). The hash package only exposes interfaces.
+
+```go
+package "hash"
+
+type Hash interface {
+  ...
+}
+
+type Hash32 interface {
+	Hash
+	Sum32() uint32
+}
+
+type Hash64 interface {
+	Hash
+	Sum64() uint64
+}
+```
+
+I suspect the benefits of moving interfaces to a separate package instead of exposing them in the subdirectories are two-fold:
+
+1. A better namespace for the interfaces. `hash.Hash` is easier to understand than `adler32.Hash`.
+2. Standardizing how to implement a functionality. A separate package with only interfaces hints that hash functions should have the methods required by the `hash.Hash` interface.
+
+Another package with only interfaces is [`encoding`](https://golang.org/pkg/encoding/).
+
+```go
+package "encoding"
+
+type BinaryMarshaler interface {
+    MarshalBinary() (data []byte, err error)
+}
+
+type BinaryUnmarshaler interface {
+    UnmarshalBinary(data []byte) error
+}
+
+type TextMarshaler interface {
+    MarshalText() (text []byte, err error)
+}
+
+type TextUnmarshaler interface {
+    UnmarshalText(text []byte) error
+}
+```
+
+Unlike `hash`, there are no functions in the std lib that accept or return an `encoding` interface. So why is it exposed? I believe it's because they want to hint to developers a standard method signature for (un)marshalling a binary into an object. Existing packages that test if a value implements the `encoding.BinaryUnmarshaler` interface won't need to change their implementation if a new struct implements the interface.
```

Comment:
I find this paragraph confusing because the [Overview](https://golang.org/pkg/encoding/#pkg-overview) section seems to indicate these exported interfaces are used in the standard library.

Reply:
ah the standard lib doesn't have any functions that accept or return an `encoding` interface. But it has many structs that implements the interface.

The only time it's used in the std lib is like this:
```go
data, err = v.Interface().(encoding.BinaryMarshaler).MarshalBinary()
```

Reply:
I updated the post like this:

There are many structs in the std lib that implement the `encoding` interfaces. However, unlike `hash` which is consumed by the packages under `crypto/`, there are no functions in the std lib that accept or return an `encoding` interface. So why is it exposed?  
I believe it's because they want to hint to developers a standard method signature for (un)marshalling a binary into an object. Existing packages that test if a value implements the `encoding.BinaryMarshaler` interface won't need to change their implementation if a new struct implements the interface.
```go
if m, ok := v.(encoding.BinaryMarshaler); ok {
    return m.MarshalBinary()
}
```